src/libraries/System.Private.CoreLib/src/System/IO/FileStatus.Unix.cs

```diff
@@ -399,17 +401,68 @@ internal long GetLength(ReadOnlySpan<char> path, bool continueOnError = false)
             return EntryExists ? _fileCache.Size : 0;
         }
 
+        internal UnixFileMode GetUnixFileMode(ReadOnlySpan<char> path, bool continueOnError = false)
+            => GetUnixFileMode(handle: null, path, continueOnError);
+
+        internal UnixFileMode GetUnixFileMode(SafeFileHandle handle, bool continueOnError = false)
+            => GetUnixFileMode(handle, handle.Path, continueOnError);
+
+        private UnixFileMode GetUnixFileMode(SafeFileHandle? handle, ReadOnlySpan<char> path, bool continueOnError = false)
+        {
+            EnsureCachesInitialized(handle, path, continueOnError);
+
+            if (!EntryExists)
+                return (UnixFileMode)(-1);
+
+            return (UnixFileMode)(_fileCache.Mode & (int)FileSystem.ValidUnixFileModes);
+        }
+
+        internal void SetUnixFileMode(string path, UnixFileMode mode)
+            => SetUnixFileMode(handle: null, path, mode);
+
+        internal void SetUnixFileMode(SafeFileHandle handle, UnixFileMode mode)
+            => SetUnixFileMode(handle, handle.Path, mode);
+
+        private void SetUnixFileMode(SafeFileHandle? handle, string? path, UnixFileMode mode)
+        {
+            if ((mode & ~FileSystem.ValidUnixFileModes) != 0)
+            {
+                throw new ArgumentException(SR.Arg_InvalidUnixFileMode, nameof(UnixFileMode));
+            }
+
+            EnsureCachesInitialized(handle, path);
+
+            if (!EntryExists)
+                FileSystemInfo.ThrowNotFound(path);
+
+            int rv = handle is not null ? Interop.Sys.FChMod(handle, (int)mode) :
+                                          Interop.Sys.LChMod(path!, (int)mode);
```

Comment:
In cast the item is a link, should we change the link permissions (`LChMod`), or its target (`ChMod`)?
Currently it does the first.

OSX supports changing link permissions, and Linux does not.

It would be most practical to let this set the target permissions. Though that gives away the option to set permissions of a link on OSX.

Probably we'd like the getter to match whatever the setter is controlling?

My preference is to make get/set operate on the link target.

@eerhardt @Jozkee @danmoseley @stephentoub what makes sense to you?

Reply:
Operating on target would make Linux/macOS consistent, but would require a new API to operate on links?

@Jozkee @carlossanlop will have perspective here from doing the symlinks API.

Reply:
CI results are interesting. On my Fedora machine, the tests pass, while in CI they don't and throw `System.IO.IOException : Operation not supported` even when setting the filemode on a file that is not a link.
Probably the libc/kernel unconditionally return the error on those system when `AT_SYMLINK_NOFOLLOW` is specified.

So if we'd like the semantics to be: target the link itself. Then the Linux implementation needs to determine the type. When it's a link, it should throw not supported, and when it is not a link it will perform a `ChMod`.

Reply:
@tmds My opinion is that we should match the behavior of other setter methods in this same file.

For example, `SetAttributes` calls `Interop.Sys.LChflags`, which sets the attributes on the link itself.

> _Though that gives away the option to set permissions of a link on OSX._

Agreed, the user should be able to decide if they want to change the link or the target's properties.

Note: In case any specific OS does not allow setting the mode on the link, then fallback to the target.

> _Probably we'd like the getter to match whatever the setter is controlling?_

That would also be my expectation.

For example: in this same file, `GetAttributes` will retrieve the `HasSymbolicLinkFlag` attribute for the link itself.

Reply:
> Note: In case any specific OS does not allow setting the mode on the link, then fallback to the target.

(As mentioned, Linux doesn't support it.)

This fallback makes the effect of the API platform dependent for links. Do we want that?

Reply:
The 3 options are, with their downside:

- get/set operates on the link.
Linux doesn't have link permissions. set is not supported (so it throws), get always returns `lrwxrwxrwx`.
- get/set operates on the target of the link.
On BSD/OSX: it is no longer possible to get/set link permissions using this API.
- get/set operates on the link if the platform supports it (BSD/OSX); and on the target of the link if the platform doesn't (Linux).
The API behaves different depending on the platform.

Reply:
@carlossanlop @dotnet/area-system-io @jeffhandley @danmoseley @stephentoub can you decide what is the desired behavior?

Reply:
My opinion (for what it counts, which is probably not very much since I'm no File System expert) is that I wouldn't have different OSes behave differently. If we think get/set permissions on the link is important, I would think adding a new API (just like how FreeBSD added `lchmod`) makes the most sense. (Note it could just be a `bool` parameter on `File.GetUnixFileMode/SetUnixFileMode`.)  Then Linux can throw PNSE if the caller tries calling the new "Don't follow symbolic links" API.

I think this follows the "principle of least surprise".

Reply:
> For example, SetAttributes calls Interop.Sys.LChflags, which sets the attributes on the link itself.

I guess if the `FileSystemInfo.Attributes` already behaves differently depending on the OS, then maybe that is a good enough reason to behave the same in UnixFileMode. Consistency in our APIs.

Reply:
> I guess if the FileSystemInfo.Attributes already behaves differently depending on the OS, then maybe that is a good enough reason to behave the same in UnixFileMode. Consistency in our APIs.

This behaves the same on all OSes.

In general, it affects the link.
For `FileAttributes.ReadOnly` we operate on the target:
https://github.com/dotnet/runtime/blob/24bbfbfb925d1d71d6c2f52d521aab00c65446a9/src/libraries/System.Private.CoreLib/src/System/IO/FileStatus.Unix.cs#L256-L260

>get/set operates on the target of the link.
On BSD/OSX: it is no longer possible to get/set link permissions using this API.

To this, I should have added: the use-cases for setting link permissions seem limited/non-existing.

My preference is to operate on the target permissions. That does something sensible on all platforms. And matches the behavior of the `chmod` user command.

Reply:
> My preference is to operate on the target permissions. That does something sensible on all platforms. And matches the behavior of the chmod user command.

+1.

It would also be good to check what Java does, just to add one more data point.

Reply:
OK, then I think we have our answer (unless someone can chime in with a really good reason not to use this behavior).